File: macos/ArmyTaskLauncher.m
#import <Cocoa/Cocoa.h>
#include <fcntl.h>
#include <limits.h>
#include <mach-o/dyld.h>
#include <spawn.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/stat.h>
#include <unistd.h>

extern char **environ;

static int startTask(void) {
  char executablePath[PATH_MAX];
  uint32_t executablePathSize = sizeof(executablePath);

  if (_NSGetExecutablePath(
        executablePath,
        &executablePathSize
      ) != 0) {
    return 1;
  }

  char *lastSlash = strrchr(executablePath, '/');

  if (lastSlash == NULL) {
    return 1;
  }

  *lastSlash = '\0';

  char appBundlePath[PATH_MAX];
  snprintf(
    appBundlePath,
    sizeof(appBundlePath),
    "%s/../..",
    executablePath
  );

  char resolvedAppBundlePath[PATH_MAX];

  if (realpath(
        appBundlePath,
        resolvedAppBundlePath
      ) == NULL) {
    return 1;
  }

  char dataDirectory[PATH_MAX];
  snprintf(
    dataDirectory,
    sizeof(dataDirectory),
    "%s",
    resolvedAppBundlePath
  );

  lastSlash = strrchr(dataDirectory, '/');

  if (lastSlash == NULL) {
    return 1;
  }

  *lastSlash = '\0';

  char binaryPath[PATH_MAX];
  snprintf(
    binaryPath,
    sizeof(binaryPath),
    "%s/Contents/Resources/army-task-bin",
    resolvedAppBundlePath
  );

  char launchLogPath[PATH_MAX];
  snprintf(
    launchLogPath,
    sizeof(launchLogPath),
    "%s/ArmyTask-launch.log",
    dataDirectory
  );

  if (setenv(
        "ARMY_TASK_DATA_DIR",
        dataDirectory,
        1
      ) != 0) {
    return 1;
  }

  int launchLogFd = open(
    launchLogPath,
    O_WRONLY | O_CREAT | O_APPEND,
    0644
  );

  if (launchLogFd < 0) {
    return 1;
  }

  posix_spawn_file_actions_t fileActions;
  posix_spawn_file_actions_init(&fileActions);
  posix_spawn_file_actions_adddup2(
    &fileActions,
    launchLogFd,
    STDOUT_FILENO
  );
  posix_spawn_file_actions_adddup2(
    &fileActions,
    launchLogFd,
    STDERR_FILENO
  );
  posix_spawn_file_actions_addclose(
    &fileActions,
    launchLogFd
  );

  char *childArgv[] = { binaryPath, NULL };
  pid_t childPid;
  int spawnResult = posix_spawn(
    &childPid,
    binaryPath,
    &fileActions,
    NULL,
    childArgv,
    environ
  );

  posix_spawn_file_actions_destroy(&fileActions);
  close(launchLogFd);

  return spawnResult == 0 ? 0 : 1;
}

int main(void) {
  @autoreleasepool {
    if (startTask() != 0) {
      return 1;
    }

    NSApplication *application = [NSApplication sharedApplication];
    [application setActivationPolicy:NSApplicationActivationPolicyAccessory];
    [application run];
  }

  return 0;
}

File: macos/ArmyTaskLauncher.c
#include <fcntl.h>
#include <limits.h>
#include <mach-o/dyld.h>
#include <spawn.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/stat.h>
#include <unistd.h>

extern char **environ;

int main(void) {
  char executable_path[PATH_MAX];
  uint32_t executable_path_size = sizeof(executable_path);

  if (_NSGetExecutablePath(
        executable_path,
        &executable_path_size
      ) != 0) {
    return 1;
  }

  char *last_slash = strrchr(executable_path, '/');

  if (last_slash == NULL) {
    return 1;
  }

  *last_slash = '\0';

  char app_bundle_path[PATH_MAX];
  snprintf(
    app_bundle_path,
    sizeof(app_bundle_path),
    "%s/../..",
    executable_path
  );

  char resolved_app_bundle_path[PATH_MAX];

  if (realpath(
        app_bundle_path,
        resolved_app_bundle_path
      ) == NULL) {
    return 1;
  }

  char data_directory[PATH_MAX];
  snprintf(
    data_directory,
    sizeof(data_directory),
    "%s",
    resolved_app_bundle_path
  );

  last_slash = strrchr(data_directory, '/');

  if (last_slash == NULL) {
    return 1;
  }

  *last_slash = '\0';

  char binary_path[PATH_MAX];
  snprintf(
    binary_path,
    sizeof(binary_path),
    "%s/Contents/Resources/army-task-bin",
    resolved_app_bundle_path
  );

  char launch_log_path[PATH_MAX];
  snprintf(
    launch_log_path,
    sizeof(launch_log_path),
    "%s/ArmyTask-launch.log",
    data_directory
  );

  if (setenv(
        "ARMY_TASK_DATA_DIR",
        data_directory,
        1
      ) != 0) {
    return 1;
  }

  int launch_log_fd = open(
    launch_log_path,
    O_WRONLY | O_CREAT | O_APPEND,
    0644
  );

  if (launch_log_fd < 0) {
    return 1;
  }

  posix_spawn_file_actions_t file_actions;
  posix_spawn_file_actions_init(&file_actions);
  posix_spawn_file_actions_adddup2(
    &file_actions,
    launch_log_fd,
    STDOUT_FILENO
  );
  posix_spawn_file_actions_adddup2(
    &file_actions,
    launch_log_fd,
    STDERR_FILENO
  );
  posix_spawn_file_actions_addclose(
    &file_actions,
    launch_log_fd
  );

  char *child_argv[] = { binary_path, NULL };
  pid_t child_pid;
  int spawn_result = posix_spawn(
    &child_pid,
    binary_path,
    &file_actions,
    NULL,
    child_argv,
    environ
  );

  posix_spawn_file_actions_destroy(&file_actions);
  close(launch_log_fd);

  return spawn_result == 0 ? 0 : 1;
}
